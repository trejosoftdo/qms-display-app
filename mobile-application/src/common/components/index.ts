export { default as ActionButton } from './ActionButton';
export { default as ToConfigurationNavigationButton } from './ToConfigurationNavigationButton';
export { default as AppView } from './AppView';
export { default as Value } from './Value';
export { default as ConditionalContainer } from './ConditionalContainer';
export { default as Cards } from './Cards';
export { default as AppHeader } from './AppHeader';
