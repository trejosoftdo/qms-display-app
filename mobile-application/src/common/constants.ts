export const AUTHORIZATION_PENDING_CODE = 'authorization_pending';

export const VALIDATION_ERROR_STATUS_CODE = 400;

export const DEVICE_NOT_CONNECTED_ERROR_MESSAGE = 'Device is not connected';

export const CONNECTION_EXPIRED_ERROR_MESSAGE = 'Connection expired!';

export const APP_PROTO = 'app://';

export const DEVICE_ID_KEY = 'SECURE_DEVICE_ID';

export const CONNECTION_DETAILS_KEY = 'CONNECTION_DETAILS';

export const WEB_PLATFORM = 'web';

export const BEARER_PORTION = 'Bearer';

export const EMPTY_VALUE = '';

export const DEFAULT_REGULAR_FONT_NAME = 'Montserrat-Regular';

export const DEFAULT_LIGHT_FONT_NAME = 'Montserrat-Light';

export const DEFAULT_MEDIUM_FONT_NAME = 'Montserrat-Medium';

export const NORMAL_FONT_WEIGHT = 'normal';

export const ONE_SECOND_MILISECONDS = 1000;

export const INTERVAL_TIME = 10 * ONE_SECOND_MILISECONDS;

export const SERVICES_PATH = '/services';

export const TICKET_DETAILS_PATH = '/ticket-details';

export const HOME_PATH = '/';

export const CATEGORIES_PATH = '/categories';

export const CONFIGURATION_PATH = '/configuration';

export const NO_OP = () => {};


export const LANGUAGES = {
  SPANISH: 'es',
  ENGLISH: 'en',
};

export const CONFIG_OPTIONS = {
  INFORMATION: 'information',
  CONNECTION: 'connection',
  LANGUAGE: 'language',
};
