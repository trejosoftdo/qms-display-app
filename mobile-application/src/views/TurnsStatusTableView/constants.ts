export const BEING_ATTENDED_STATUS_CODE = "BEING_ATTENDED";

export const TO_BE_ATTENDED_STATUS_CODE = "TO_BE_ATTENDED";