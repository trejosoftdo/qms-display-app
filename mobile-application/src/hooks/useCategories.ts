import { useProgress } from "../common/hooks";
import { CategoriesData, Progress } from "../common/models";
import { loadCategories } from "../common/services/category";

/**
 * Hook to get the available categories
 * @returns Progress<CategoriesData>
 */
const useCategories = (): Progress<CategoriesData> => useProgress<CategoriesData>(loadCategories());

export default useCategories;
