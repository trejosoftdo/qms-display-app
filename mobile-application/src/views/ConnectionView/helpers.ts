import { ConnectionData, ConnectionDetails } from "../../common/models";
import { TRANSLATION_CONNECTED_MESSAGE_KEY, TRANSLATION_INITIAL_CONNECT_MESSAGE_KEY, TRANSLATION_SUCCESS_CONNECTION_MESSAGE_KEY, TRANSLATION_TO_CONNECT_MESSAGE_KEY } from "../../common/translations/translation-keys";

/**
 * Gets the correct message key from the data state
 * 
 * @param  {ConnectionData} data
 * @returns string
 */
export const getMessageKey = (data: ConnectionData, connectionDetails?: ConnectionDetails): string => {
  if (connectionDetails?.accessToken) {
    return TRANSLATION_CONNECTED_MESSAGE_KEY;
  }

  if (data?.tokens?.accessToken) {
    return TRANSLATION_SUCCESS_CONNECTION_MESSAGE_KEY;
  }

  return data?.userCode ? TRANSLATION_TO_CONNECT_MESSAGE_KEY : TRANSLATION_INITIAL_CONNECT_MESSAGE_KEY;
};
