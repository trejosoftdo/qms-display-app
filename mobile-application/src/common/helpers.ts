import { Platform } from 'react-native';
import { router } from 'expo-router';
import { APP_PROTO, EMPTY_VALUE, ONE_SECOND_MILISECONDS, WEB_PLATFORM } from './constants';

/** 
 * Checks if the platform is web
 * @returns boolean
 */
export const isWeb = (): boolean => Platform.OS === WEB_PLATFORM;

/** 
 * Gets the current time
 * @returns number
 */
export const getCurrentTime = (): number => new Date().getTime();

/**
 * Calculates the expire time
 * 
 * @param  {number} expiresIn
 * @returns number
 */
export const calculateExpireTime = (expiresIn: number): number => getCurrentTime() + expiresIn * ONE_SECOND_MILISECONDS;


/** 
 * Extracs icon name
 * @returns string
 */
export const extractIconName = (iconUrl: string): string => iconUrl.replace(APP_PROTO, EMPTY_VALUE)

/**
 * Navigates to a given path
 * 
 * @param  {string} pathname
 * @param  {any} params
 * @returns void
 */
export const goToPath = (pathname: string, params: Record<string, any> = undefined): void => {
  router.push({
    pathname,
    params,
  });
};

/**
 * Waits for certain time
 * @param  {number} timeout 
 * @returns Promise<any>
 */
export const delay = (timeout: number) => new Promise((resolve) => {
  setTimeout(resolve, timeout);
});

/**
 * Gets the response blob url
 * @param  {Response} response 
 * @returns Promise<string>
 */
export const getResponseBlobUrl = async (response: Response): Promise<string> => {
  const stream = await response.blob();
  return URL.createObjectURL(stream);
};
