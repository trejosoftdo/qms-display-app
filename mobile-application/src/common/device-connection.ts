import { v4 as uuidv4 } from 'uuid';
import { AuthHeaders, ConnectionDetails } from './models';
import { CONNECTION_DETAILS_KEY, DEVICE_ID_KEY } from './constants';
import { DEVICE_NOT_CONNECTED_ERROR, CONNECTION_EXPIRED_ERROR } from './errors';
import { getValue, setValue } from './store';
import { getNewAccessToken } from './services/auth';
import { BEARER_PORTION } from './constants';
import { getCurrentTime } from './helpers';


/**
 * Checks if a time has expired
 * 
 * @param  {{expiresAt:number}} {expiresAt}
 * @returns boolean
 */
const hasExpired = ({ expiresAt }: { expiresAt: number }): boolean => getCurrentTime() > expiresAt;

/**
 * Gets the device/installation id
 * @returns Promise<string>
 */
export const getDeviceId = async (): Promise<string> => {
  const fetchUUID = await getValue(DEVICE_ID_KEY);
  const deviceId = fetchUUID || uuidv4();
  await setValue(DEVICE_ID_KEY, deviceId.toString());
  return deviceId;
};

/**
 * Saves the connection details
 * 
 * @param  {ConnectionDetails} connectionDetails
 * @returns Promise<void>
 */
export const saveConnectionDetails = async (connectionDetails: ConnectionDetails): Promise<void> => {
  await setValue(CONNECTION_DETAILS_KEY, JSON.stringify(connectionDetails));
};

/**
 * Gets the stored connection details
 * 
 * @returns Promise<ConnectionDetails | undefined>
 */
export const getConnectionDetails = async (): Promise<ConnectionDetails | undefined> => {
  try {
    const data = await getValue(CONNECTION_DETAILS_KEY);
    const details = JSON.parse(data);
    return details as ConnectionDetails;
  } catch (error) {
    return undefined;
  }
};

/**
 * Gets the device auth headers to consume the APIs
 * 
 * @returns AuthHeaders
 */
export const getDeviceAuthHeaders = async (): Promise<AuthHeaders> => {
  const connectionDetails = await getConnectionDetails();

  if (!(connectionDetails?.applicationId && connectionDetails?.accessToken)) {
    throw DEVICE_NOT_CONNECTED_ERROR;
  }

  if (hasExpired(connectionDetails.accessToken)) {
    if (hasExpired(connectionDetails.refreshToken)) {
      throw CONNECTION_EXPIRED_ERROR;
    }

    connectionDetails.accessToken = await getNewAccessToken(
      connectionDetails.applicationId,
      connectionDetails.refreshToken.value,
    );
    await saveConnectionDetails(connectionDetails);
  }

  return {
    applicationId: connectionDetails.applicationId,
    authorization: `${BEARER_PORTION} ${connectionDetails.accessToken.value}`,
  };
};
