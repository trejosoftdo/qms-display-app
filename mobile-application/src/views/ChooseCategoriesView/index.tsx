import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocalSearchParams } from 'expo-router';
import { AppView, Cards, ConditionalContainer } from '../../common/components';
import { goToPath } from '../../common/helpers';
import useCategories from '../../hooks/useCategories';
import { SERVICES_PATH } from '../../common/constants';
import { TRANSLATION_CHOOSE_CATEGORY_MESSAGE_KEY, TRANSLATION_WAIT_MESSAGE_KEY } from '../../common/translations/translation-keys';


/**
 * ChooseCategoriesViewProps defines the props for the Choose Categories View Component.
 */
interface ChooseCategoriesViewProps {}

/**
 * A component that represents the view for choosing categories
 *
 * @param {ChooseCategoriesViewProps} props - The props for the Choose Categories View component.
 */
const ChooseCategoriesView: React.FC<ChooseCategoriesViewProps> = (props: ChooseCategoriesViewProps) => {
  const params = useLocalSearchParams();
  const {
    loading,
    data,
    error,
  } = useCategories();
  const { t } = useTranslation();
  const messageKey = loading ? TRANSLATION_WAIT_MESSAGE_KEY : TRANSLATION_CHOOSE_CATEGORY_MESSAGE_KEY;
  return (
    <AppView
      headerMessage={t(messageKey)}
      loading={loading}
      error={error}
    >
      <ConditionalContainer display={!loading && !!data}>
        <Cards
          items={data?.items}
          onItemSelect={(item) => {
            goToPath(
              SERVICES_PATH,
              {
                customerName: params.customerName,
                categoryId: item.id,
                categoryLabel: item.label,
              },
            );
          }}
        />
      </ConditionalContainer>
    </AppView>
  );
};

export default ChooseCategoriesView;