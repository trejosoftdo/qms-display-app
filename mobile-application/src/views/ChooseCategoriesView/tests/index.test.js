import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { useLocalSearchParams } from 'expo-router';
import { goToPath } from '../../../common/helpers';
import { SERVICES_PATH } from '../../../common/constants';
import useCategories from '../../../hooks/useCategories';
import ChooseCategoriesView from '../';


jest.mock('expo-router');
jest.mock('../../../common/helpers');
jest.mock('../../../hooks/useCategories');


describe('ChooseCategoriesView component', () => {
  const mockParams = {
    customerName: 'customer-name',
  };
  const mockData = {
    items: [{
      id: 'mock-id',
      name: 'mock-name',
      label: 'mock-label',
    }],
  };

  beforeEach(() => {
    useLocalSearchParams.mockReturnValue(mockParams);
  });

  afterEach(() => {
    jest.clearAllMocks();
  })

  it('renders the component as expected when loading', () => {
    useCategories.mockReturnValue({ loading: true });
    expect(renderShallow(<ChooseCategoriesView />)).toMatchSnapshot();
    expect(useLocalSearchParams).toHaveBeenCalledTimes(1);
    expect(useLocalSearchParams).toHaveBeenCalledWith();
    expect(useCategories).toHaveBeenCalledTimes(1);
    expect(useCategories).toHaveBeenCalledWith();
  });

  it('renders the component as expected when data has been loaded', () => {
    useCategories.mockReturnValue({
      loading: false,
      data: mockData,
    });
    expect(renderShallow(<ChooseCategoriesView />)).toMatchSnapshot();
    expect(useLocalSearchParams).toHaveBeenCalledTimes(1);
    expect(useLocalSearchParams).toHaveBeenCalledWith();
    expect(useCategories).toHaveBeenCalledTimes(1);
    expect(useCategories).toHaveBeenCalledWith();
  });

  it('renders the component as expected when it has errors', () => {
    useCategories.mockReturnValue({
      loading: false,
      error: new Error('Unexpected error'),
    });
    expect(renderShallow(<ChooseCategoriesView />)).toMatchSnapshot();
    expect(useLocalSearchParams).toHaveBeenCalledTimes(1);
    expect(useLocalSearchParams).toHaveBeenCalledWith();
    expect(useCategories).toHaveBeenCalledTimes(1);
    expect(useCategories).toHaveBeenCalledWith();
  });

  it('goes to services path when an item is selected', () => {
    useCategories.mockReturnValue({
      loading: false,
      data: mockData,
    });
    const { getByText } = render(<ChooseCategoriesView />);
    const element = getByText('mock-label');

    fireEvent.press(element);

    expect(goToPath).toHaveBeenCalledTimes(1);
    expect(goToPath).toHaveBeenCalledWith(
      SERVICES_PATH,
      {
        categoryId: 'mock-id',
        categoryLabel: 'mock-label',
        customerName: 'customer-name'
      },
    );
    expect(useLocalSearchParams).toHaveBeenCalledTimes(1);
    expect(useLocalSearchParams).toHaveBeenCalledWith();
    expect(useCategories).toHaveBeenCalledTimes(1);
    expect(useCategories).toHaveBeenCalledWith();
  });
});
