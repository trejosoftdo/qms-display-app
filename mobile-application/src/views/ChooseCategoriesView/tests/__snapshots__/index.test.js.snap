// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChooseCategoriesView component renders the component as expected when data has been loaded 1`] = `
<AppView
  headerMessage="Translated[translation:chooseCategoryMessage]({})"
  loading={false}
>
  <ConditionalContainer
    display={true}
  >
    <Cards
      items={
        [
          {
            "id": "mock-id",
            "label": "mock-label",
            "name": "mock-name",
          },
        ]
      }
      onItemSelect={[Function]}
    />
  </ConditionalContainer>
</AppView>
`;

exports[`ChooseCategoriesView component renders the component as expected when it has errors 1`] = `
<AppView
  error={[Error: Unexpected error]}
  headerMessage="Translated[translation:chooseCategoryMessage]({})"
  loading={false}
>
  <ConditionalContainer
    display={false}
  >
    <Cards
      onItemSelect={[Function]}
    />
  </ConditionalContainer>
</AppView>
`;

exports[`ChooseCategoriesView component renders the component as expected when loading 1`] = `
<AppView
  headerMessage="Translated[translation:waitMessage]({})"
  loading={true}
>
  <ConditionalContainer
    display={false}
  >
    <Cards
      onItemSelect={[Function]}
    />
  </ConditionalContainer>
</AppView>
`;
